### ISSUE TITLE
Optimizing Git Artifact Handling

### ISSUE THREAD

**Issue by mtiller** [argoproj/argo#3254]
# Summary

I'm a very enthusiastic fan of `argo`.  It takes care of so many useful things.  However, I'm wondering if there has been any discussion around optimizing the handling of `git` artifacts?  I looked through the issues (open and closed) and didn't really see any mention specifically how `argo` could cache `git` repos in the cluster to improve performance.

# Motivation

When using `git` artifacts from `git` repos outside the cluster (e.g., Github), the clone process can be quite time consuming and you can even run into rate limiting issues.  In practice, you can quickly get to a point where the vast majority of information traveling over the wire is repeated.

It strikes me that one could use a PVC as a place to mirror `git` repos.  In this way, the use of a `git` artifact would first be to sync the mirror and then perform a checkout by mounting the mirror read only into the container.  The initial "sync" of this mirror would be effectively involve the same time and data as the current approach.  But the subsequent sync operations would simply involve one or more fetch operations to ensure things were in sync which is far less data.  Keeping the mirror cluster local just seems like a huge potential efficiency gain.

Now, this clearly would require some development effort.  I guess my main question is whether this has been discussed before and whether any faults were identified?  One obvious issue that occurs to me would be around security.  But since `git` repos don't have fine grained authorization machinery, the only real question is whether a given workflow should have access to the mirror or not and that can be answered by whether the "sync" fetching is successful (if so, there is no reason not to give the user access to the entire local `git` mirror).

But I could easily be overlooking something, so feel free to tell me why this is a stupid idea. ;-)

**Comment by alexec** [comment]
To improve performance do two things:

1. Shallow clone using `depth: 1`.
2. Clone once to a volume [as per the the CI example](https://github.com/argoproj/argo/blob/master/examples/ci.yaml).

**Comment by mtiller** [comment]
Shallow clone does help.  But we have use cases where we aren't necessarily working on a branch tip.  We'd like the flexibility to clone an arbitrary commit (may not be a branch or even tagged) and [that isn't really possible](https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1).

As for the CI case, it looks to me from that example that it is creating a temporary PVC *for this individual workflow* and then reusing it across steps.  That's fine.  But it doesn't really address my concern which is that we run lots of workflows that share the same `git` repo.  So we can't cache the result within a workflow.

I completely recognize that you could take care of all this within your workflow.  You could have the first step of the workflow be to trigger a syncing of the repo and then mount the synced `git` repo read only and locally, much as I've described.  So I realize it can be done in `argo`.  It just seems like, from the user's perspective, such mirroring/caching could provide benefits for nearly anybody using `git` artifacts vs. having to roll their own solution one user at a time.

But talk it cheap, I realize it would be a lot of work to implement.  I'm just trying to figure out whether the concept itself is flawed in some way I can't see or if it is *only* the effort required that has prevented people from implementing it.

**Comment by alexec** [comment]
You're correct. We could expose the `git clone --depth 1 -b branch ...` - but this does not work for tags or hashes.
Could you run a Git proxy in your cluster?

**Comment by mtiller** [comment]
I'm not sure what you mean by a "git proxy".  If I search for "git proxy", all I get are articles about running git behind an http proxy.  Is there a specific tool you were thinking of here, because I've been unable to find something that does what I describe "out of the box".

**Comment by mtiller** [comment]
I did come across [this](https://stackoverflow.com/questions/32194022/local-cache-for-a-github-repository).  I'll look into it more closely tomorrow.

**Comment by alexec** [comment]
https://github.com/google/goblet?

**Comment by mtiller** [comment]
I saw that.  The repo has exactly one commit.  It doesn't really inspire a lot of confidence.  I also looked a bit more into [`git-cache-http-server`](https://github.com/jonasmalacofilho/git-cache-http-server).  I'm giving that a go.  Definitely useful, but doesn't quite solve the whole problem. With `git-cache-http-server`, I can keep a cluster local mirror (effectively).

I can't speak for others, but most of the time in my workflows what I really want in my workflow is to do a *checkout*, not a clone/fetch of a specific commit (not necessarily a branch or tag).  The depth settings, etc. kind of mimic this, but they don't work for commits in general.  So what I really need is not just a local server that optimizes bandwidth (which is what `git-cache-httpserver` helps with), but I'd like to actually have a read-only repo directly accessible from the container so that I can simply checkout just the files I need (vs. replicating a whole repo just for the sake of a few files).  As far as I can tell, argo always does a clone/fetch followed by a checkout.  This is a lot more data than I really need for my workflow.  So having a read-only repo already mounted to my container would allow for just a checkout which is really just what I need.

Even better would be a cache of the last checkout so I could just refresh that checkout (because the files won't change that much).  But that may be asking a bit much and introduce a bunch of "statefulness" that could lead to odd errors related to side effect.
